Fix modbus default delay

janiversen opened the pull request:
<!--
  You are amazing! Thanks for contributing to our project!
  Please, DO NOT DELETE ANY TEXT from this template! (unless instructed).
-->
## Breaking change
<!--
  If your PR contains a breaking change for existing users, it is important
  to tell them what breaks, how to make it work again and why we did this.
  This piece of text is published with the release notes, so it helps if you
  write it towards our users, not us.
  Note: Remove this section if this PR is NOT a breaking change.
-->


## Proposed change
<!-- 
  Describe the big picture of your changes here to communicate to the
  maintainers why we should accept this pull request. If it fixes a bug
  or resolves a feature request, be sure to link to that issue in the
  additional information section.
-->

CONF_DELAY was used in a serial connection, which is not
permitted.

Sometimes async_update is called after async_setup is completed,
but before event EVENT_HOMEASSISTANT_START is issued, leading to
a missing object. So added a test for missing object.

## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box! 
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [ x] Bugfix (non-breaking change which fixes an issue)
- [ ] New integration (thank you!)
- [ ] New feature (which adds functionality to an existing integration)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Example entry for `configuration.yaml`:
<!--
  Supplying a configuration snippet, makes it easier for a maintainer to test
  your PR. Furthermore, for new integrations, it gives an impression of how
  the configuration would look like.
  Note: Remove this section if this PR does not have an example entry.
-->

```yaml
# Example configuration.yaml

```

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

- This PR fixes or closes issue: fixes #
- This PR is related to issue: 
#33873
- Link to documentation pull request: 
https://github.com/home-assistant/home-assistant.io/pull/12762

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [ x] The code change is tested and works locally.
- [x ] Local tests pass. **Your PR cannot be merged unless tests pass**
- [ x] There is no commented out code in this PR.
- [x ] I have followed the [development checklist][dev-checklist]
- [x ] The code has been formatted using Black (`black --fast homeassistant tests`)
- [ ] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [ ] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [ ] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [ ] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [ ] Untested files have been added to `.coveragerc`.

The integration reached or maintains the following [Integration Quality Scale][quality-scale]:
<!--
  The Integration Quality Scale scores an integration on the code quality
  and user experience. Each level of the quality scale consists of a list
  of requirements. We highly recommend getting your integration scored!
-->

- [ ] No score or internal
- [ ] 🥈 Silver
- [ ] 🥇 Gold
- [ ] 🏆 Platinum

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/en/development_checklist.html
[manifest-docs]: https://developers.home-assistant.io/docs/en/creating_integration_manifest.html
[quality-scale]: https://developers.home-assistant.io/docs/en/next/integration_quality_scale_index.html
[docs-repository]: https://github.com/home-assistant/home-assistant.io

janiversen committed:
solve modbus issue #33872

CONF_DELAY was used in a serial connection, which is not
permitted.

Sometimes async_update is called after async_setup is completed,
but before event EVENT_HOMEASSISTANT_START is issued, leading to
a missing object.

MartinHjelmare commented on homeassistant/components/modbus/__init__.py:
```diff
@@ -266,12 +267,16 @@ def setup(self):
 
     async def read_coils(self, unit, address, count):
         """Read coils."""
+        if self._client is None:
```
Why do we need to wait to set up the client until home assistant start event?

janiversen replied:
We do not, this part of the code is before my time, and I thought it was mandatory to wait for the start event so I simply continued along the same line.

I will change that now, thanks for the advice, slowly I learn more about how the core works, so I can start helping others.

MartinHjelmare replied:
Usually the reason to wait until home assistant start event is to allow other integrations that we rely on to set up. This is common in eg template integrations that take their input from other integrations.

janiversen replied:
That is a logical explanation, but also a very crude way of doing it, which I presume cannot hold in the long term:
think a an integration depending on another integration and being used in a template integration.

Having a method wait_for_integration_to_complete(<integration>) would be a more direct way, but that is very much outside the scope of this PR.

push a new commit, with your requested changes.

MartinHjelmare replied:
Yeah, we should probably have a way to say that an integration should wait until final call before boarding so to speak, during set up but still before start event.

janiversen committed:
resolve review comment.

Do not wait for start event, but activate pymodbus directly in
async setup.

MartinHjelmare commented on homeassistant/components/modbus/__init__.py:
```diff
@@ -150,8 +149,8 @@ def start_modbus(event):
-    hass.bus.async_listen_once(EVENT_HOMEASSISTANT_START, start_modbus)
-
+    # do not wait for EVENT_HOMEASSISTANT_START, activate pymodbus now
+    start_modbus()
```
Is this call async safe?

We're mixing sync and async api inside `start_modbus`. We should select either of those APIs depending on where it should run, inside event loop or outside.

janiversen replied:
yeah I am painfully aware of that. Problem is that pymodbus needs a sync setup call, while all of our code is async. I will isolate the setup call in sync shield.

MartinHjelmare replied:
We can use the executor if it should run in another thread. Use `hass.async_add_executor_job` for that.

janiversen replied:
I am trying exactly that, but for some reason the setup hangs, so obviously I have done something wrong.

MartinHjelmare replied:
Push what you have and I can take a look.

janiversen replied:
thanks. I have pushed my latest.

I think the old code was attaching to the start event to secure a pure sync call (at least I remember having seen this problem earlier, when I converted everything to async).

I

janiversen replied:
“madre mia” ! (a good spanish expression. It was hanging in listen_once, and I also needed an await (of course) in the executor line.

Now it works for me.

janiversen committed:
review 2

Remark, this does not work, async_setup hangs.

clean start_modbus() from async calls, leaving only the pymodbus
setup.

janiversen committed:
review 2a

Moved listen_once back to start_modbus, since it is sync.

MartinHjelmare reviewed (approved):
Great!

MartinHjelmare commented:
Should we close #33873 with this PR?

janiversen commented:
I would prefer to keep the issue open until I have a confirmation from the people that it works, if that is not a problem for you.

MartinHjelmare commented:
Ok.
